File: app/src/main/java/com/longkai/stcarcontrol/st_exp/activity/ChooseActivity.java
package com.longkai.stcarcontrol.st_exp.activity;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import androidx.annotation.NonNull;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.longkai.stcarcontrol.st_exp.ConstantData;
import com.longkai.stcarcontrol.st_exp.R;
import com.longkai.stcarcontrol.st_exp.STCarApplication;
import com.longkai.stcarcontrol.st_exp.Utils.SharedPreferencesUtil;
import com.longkai.stcarcontrol.st_exp.communication.ConnectionListener;
import com.longkai.stcarcontrol.st_exp.communication.ConnectionType;
import com.longkai.stcarcontrol.st_exp.communication.ServiceManager;
import com.longkai.stcarcontrol.st_exp.communication.commandList.BaseResponse;
import com.longkai.stcarcontrol.st_exp.communication.commandList.CMDGetVersion;
import com.longkai.stcarcontrol.st_exp.communication.commandList.CommandListenerAdapter;

import java.util.Timer;
import java.util.TimerTask;


/**
 *
 * Created by Administrator on 2018/5/12.
 */

public class ChooseActivity extends BaseActivity implements View.OnClickListener {

    private ImageView ivBTConnectionState, ivWifiConnectionState;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_choose_first);
        STCarApplication.verifyStoragePermissions(this);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode){
            case 1:
                for (String permission:
                     permissions) {
                    if (permission.equalsIgnoreCase(Manifest.permission.WRITE_EXTERNAL_STORAGE)){
                        STCarApplication.logConfig();
                    }
                }
                break;
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        initUI();

        String verName = "version";
        try {
            verName = getApplicationContext().getPackageManager().
                    getPackageInfo(getApplicationContext().getPackageName(), 0).versionName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        ((TextView)findViewById(R.id.tv_choose_activity_version)).setText(verName);
        ServiceManager.getInstance().init(getApplicationContext(), new ServiceManager.InitCompleteListener() {
            @Override
            public void onInitComplete() {
                ServiceManager.getInstance().setConnectionListener(mConnectionListener);
            }
        });

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.d("ChooseActivity", "onDestroy: ChooseActivity");
        ServiceManager.getInstance().destroy();
    }

    private void initUI(){
        findViewById(R.id.btn_choose_function).setOnClickListener(this);
        findViewById(R.id.btn_choose_VCU).setOnClickListener(this);
        findViewById(R.id.btn_choose_entertainment).setOnClickListener(this);
        ivBTConnectionState = (ImageView) findViewById(R.id.iv_chooseactivity_lost_bluetooth);
        ivBTConnectionState.setOnClickListener(this);
        ivWifiConnectionState = (ImageView) findViewById(R.id.iv_chooseactivity_lost_wifi);
        ivWifiConnectionState.setOnClickListener(this);

        if (communicationEstablished) {
            ivWifiConnectionState.setVisibility(View.INVISIBLE);
            ivBTConnectionState.setVisibility(View.INVISIBLE);
        }

    }

    @Override
    public void onClick(View v) {
        switch (v.getId()){
            case R.id.btn_choose_function:
                Intent intent = new Intent(this, MainActivity.class);
                startActivity(intent);
                break;
            case R.id.btn_choose_VCU:
                Intent intent2 = new Intent(this, VCUActivity.class);
                startActivity(intent2);
                break;
            case R.id.btn_choose_entertainment:
                Intent intent3 = new Intent(this, InfotainmentActivity.class);
                startActivity(intent3);
                break;
            case R.id.iv_chooseactivity_lost_wifi:
                SharedPreferencesUtil.put(this, ConstantData.CONNECTION_TYPE, "WIFI");
                if (!hardwareConnected) {
                    ServiceManager.getInstance().connectToDevice(null, mConnectionListener, ConnectionType.Wifi);
                }
                ServiceManager.getInstance().sendCommandToCar(new CMDGetVersion(), getVersionListener);
                break;
            case R.id.iv_chooseactivity_lost_bluetooth:
                SharedPreferencesUtil.put(this, ConstantData.CONNECTION_TYPE, "BT");
                if (!hardwareConnected) {
                    ServiceManager.getInstance().connectToDevice(null, mConnectionListener, ConnectionType.BT);
                }
                ServiceManager.getInstance().sendCommandToCar(new CMDGetVersion(), getVersionListener);
                break;
        }
    }

    Timer timer;
    ConnectionListener mConnectionListener = new ConnectionListener() {
        @Override
        public void onConnected() {
//            Toast.makeText(getApplicationContext(), "Bt Connected", Toast.LENGTH_SHORT).show();
            hardwareConnected = true;
            if (null == timer){
                timer = new Timer();
            }

            timer.schedule(new TimerTask() {
                @Override
                public void run() {
                    ServiceManager.getInstance().sendCommandToCar(new CMDGetVersion(), getVersionListener);
                }
            }, 2000);//等两秒发送get version command

        }

        @Override
        public void onDisconnected() {
            Toast.makeText(getApplicationContext(), "Disconnected", Toast.LENGTH_LONG).show();
            hardwareConnected = false;
            communicationEstablished = false;
            ivWifiConnectionState.setVisibility(View.VISIBLE);
            ivBTConnectionState.setVisibility(View.VISIBLE);
        }
    };


    public String mVersion;
    private CommandListenerAdapter getVersionListener = new CommandListenerAdapter(){
        @Override
        public void onSuccess(BaseResponse response) {
            super.onSuccess(response);
            //invisible View
            mVersion = ((CMDGetVersion.Response)response).getVersion();
            communicationEstablished = true;

            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    ivBTConnectionState.setVisibility(View.INVISIBLE);
                    ivWifiConnectionState.setVisibility(View.INVISIBLE);
                    Toast.makeText(getApplicationContext(),
                            "version:" + mVersion ,Toast.LENGTH_SHORT).show();
                }
            });

        }

        @Override
        public void onTimeout() {
            super.onTimeout();
        }
    };
}

File: app/src/main/java/com/longkai/stcarcontrol/st_exp/activity/ChooseActivity.kt
package com.longkai.stcarcontrol.st_exp.activity

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.longkai.stcarcontrol.st_exp.ConstantData
import com.longkai.stcarcontrol.st_exp.R
import com.longkai.stcarcontrol.st_exp.STCarApplication.Companion.logConfig
import com.longkai.stcarcontrol.st_exp.STCarApplication.Companion.verifyStoragePermissions
import com.longkai.stcarcontrol.st_exp.Utils.SharedPreferencesUtil
import com.longkai.stcarcontrol.st_exp.communication.ConnectionListener
import com.longkai.stcarcontrol.st_exp.communication.ConnectionType
import com.longkai.stcarcontrol.st_exp.communication.ServiceManager
import com.longkai.stcarcontrol.st_exp.communication.commandList.BaseResponse
import com.longkai.stcarcontrol.st_exp.communication.commandList.CMDGetVersion
import com.longkai.stcarcontrol.st_exp.communication.commandList.CommandListenerAdapter
import com.longkai.stcarcontrol.st_exp.compose.DdsActivity
import com.longkai.stcarcontrol.st_exp.pbox.TrackingActivity
import java.util.*

/**
 *
 * Created by Administrator on 2018/5/12.
 */
class ChooseActivity : BaseActivity(), View.OnClickListener {
    private var ivBTConnectionState: ImageView? = null
    private var ivWifiConnectionState: ImageView? = null
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_choose_first)
        verifyStoragePermissions(this)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            1 -> for (permission in permissions) {
                if (permission.equals(
                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        ignoreCase = true
                    )
                ) {
                    logConfig()
                }
            }
        }
    }

    override fun onStart() {
        super.onStart()
        initUI()
        var verName: String? = "version"
        try {
            verName = applicationContext.packageManager.getPackageInfo(
                applicationContext.packageName,
                0
            ).versionName
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
        (findViewById<View>(R.id.tv_choose_activity_version) as TextView).text = verName
        ServiceManager.getInstance().init(applicationContext) {
            ServiceManager.getInstance().setConnectionListener(mConnectionListener)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d("ChooseActivity", "onDestroy: ChooseActivity")
        ServiceManager.getInstance().destroy()
    }

    private fun initUI() {
        findViewById<View>(R.id.btn_choose_function).setOnClickListener(this)
        findViewById<View>(R.id.btn_choose_VCU).setOnClickListener(this)
        findViewById<View>(R.id.btn_choose_entertainment).setOnClickListener(this)
        findViewById<View>(R.id.btn_choose_dds).setOnClickListener(this)
        findViewById<View>(R.id.btn_choose_tracking).setOnClickListener(this)
        ivBTConnectionState = findViewById<View>(R.id.iv_chooseactivity_lost_bluetooth) as ImageView
        ivBTConnectionState!!.setOnClickListener(this)
        ivWifiConnectionState = findViewById<View>(R.id.iv_chooseactivity_lost_wifi) as ImageView
        ivWifiConnectionState!!.setOnClickListener(this)
        if (communicationEstablished) {
            ivWifiConnectionState!!.visibility = View.INVISIBLE
            ivBTConnectionState!!.visibility = View.INVISIBLE
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_choose_function -> {
                val intent = Intent(this, MainActivity::class.java)
                startActivity(intent)
            }
            R.id.btn_choose_VCU -> {
                val intent2 = Intent(this, VCUActivity::class.java)
                startActivity(intent2)
            }
            R.id.btn_choose_entertainment -> {
                val intent3 = Intent(this, InfotainmentActivity::class.java)
                startActivity(intent3)
            }
            R.id.btn_choose_dds -> startActivity(Intent(this, DdsActivity::class.java))
            R.id.btn_choose_tracking -> startActivity(Intent(this, TrackingActivity::class.java))
            R.id.iv_chooseactivity_lost_wifi -> {
                SharedPreferencesUtil.put(this, ConstantData.CONNECTION_TYPE, "WIFI")
                if (!hardwareConnected) {
                    ServiceManager.getInstance()
                        .connectToDevice(null, mConnectionListener, ConnectionType.Wifi)
                }
                ServiceManager.getInstance().sendCommandToCar(CMDGetVersion(), getVersionListener)
            }
            R.id.iv_chooseactivity_lost_bluetooth -> {
                SharedPreferencesUtil.put(this, ConstantData.CONNECTION_TYPE, "BT")
                if (!hardwareConnected) {
                    ServiceManager.getInstance()
                        .connectToDevice(null, mConnectionListener, ConnectionType.BT)
                }
                ServiceManager.getInstance().sendCommandToCar(CMDGetVersion(), getVersionListener)
            }
        }
    }

    var timer: Timer? = null
    var mConnectionListener: ConnectionListener = object : ConnectionListener {
        override fun onConnected() {
//            Toast.makeText(getApplicationContext(), "Bt Connected", Toast.LENGTH_SHORT).show();
            hardwareConnected = true
            if (null == timer) {
                timer = Timer()
            }
            timer!!.schedule(object : TimerTask() {
                override fun run() {
                    ServiceManager.getInstance()
                        .sendCommandToCar(CMDGetVersion(), getVersionListener)
                }
            }, 2000) //等两秒发送get version command
        }

        override fun onDisconnected() {
            Toast.makeText(applicationContext, "Disconnected", Toast.LENGTH_LONG).show()
            hardwareConnected = false
            communicationEstablished = false
            ivWifiConnectionState!!.visibility = View.VISIBLE
            ivBTConnectionState!!.visibility = View.VISIBLE
        }
    }
    var mVersion: String? = null
    private val getVersionListener: CommandListenerAdapter<*> =
        object : CommandListenerAdapter<BaseResponse>() {
            override fun onSuccess(response: BaseResponse) {
                super.onSuccess(response)
                //invisible View
                mVersion = (response as CMDGetVersion.Response).version
                communicationEstablished = true
                runOnUiThread {
                    ivBTConnectionState!!.visibility = View.INVISIBLE
                    ivWifiConnectionState!!.visibility = View.INVISIBLE
                    Toast.makeText(
                        applicationContext,
                        "version:$mVersion", Toast.LENGTH_SHORT
                    ).show()
                }
            }

            override fun onTimeout() {
                super.onTimeout()
            }
        }
}